Rubberduck.Parsing/Symbols/ModuleBodyElementDeclaration.cs

```diff
@@ -117,10 +73,78 @@ protected void AddParameters(IEnumerable<ParameterDeclaration> parameters)
             _parameters.AddRange(parameters);
         }
 
+        private Lazy<(bool IsInterfaceImplementation, ClassModuleDeclaration InterfaceImplemented)> _interface;
+
+        /// <summary>
+        /// Returns true if this member is a concrete implementation of an interface.
+        /// </summary>
+        public bool IsInterfaceImplementation => _interface != null && _interface.Value.IsInterfaceImplementation;
+
+        /// <summary>
+        /// Returns the interface that this member implements from, or null if not an implementation.
+        /// </summary>
+        public ClassModuleDeclaration InterfaceImplemented => _interface?.Value.InterfaceImplemented;
+
+        private Lazy<Declaration> _implemented;
+
+        /// <summary>
+        /// Returns the interface member that this member is a concrete implementation of, or null if not an implementation.
+        /// </summary>
+        public Declaration InterfaceMemberImplemented => _implemented?.Value;
+
+        /// <inheritdoc/>
+        public string ImplementingIdentifierName => this.ImplementingIdentifierName();
+
+        /// <summary>
+        /// Returns true if this Declaration is a concrete implementation of the passed member.
+        /// </summary>
+        /// <param name="interfaceMember">The member to test for implementation of.</param>
+        /// <returns>True if this Declaration is a concrete implementation of interfaceMember.</returns>
+        protected abstract bool Implements(ICanBeInterfaceMember interfaceMember);
+
         /// <inheritdoc/>
         public bool IsInterfaceMember => this.IsInterfaceMember();
 
         /// <inheritdoc/>
         public ClassModuleDeclaration InterfaceDeclaration => this.InterfaceDeclaration();
+
+        internal void InvalidateInterfaceCache()
+        {
+            if (!(ParentDeclaration is ClassModuleDeclaration))
+            {
+                return;
+            }
+
+            _interface = new Lazy<(bool IsInterfaceImplementation, ClassModuleDeclaration InterfaceImplemented)>(() => ResolveInterface(this), true);
+            _implemented = new Lazy<Declaration>(() => MemberImplemented(this), true);
+        }
+
+        private static (bool IsInterfaceImplementation, ClassModuleDeclaration InterfaceImplemented) ResolveInterface(ModuleBodyElementDeclaration element)
+        {
+            if (!(element.ParentDeclaration is ClassModuleDeclaration classModule))
+            {
+                return (false, null);
+            }
+
+            var identifiers = element.IdentifierName.Split('_');
+
+            if (identifiers.Length == 1)
+            {
+                return (false, null);
+            }
+
+            var supertype = string.Join("_", identifiers.Take(identifiers.Length - 1));
```

Comment:
Is this assumption generally true? As in: Is it guaranteed that only the last part of the identifier refers to the member name? If yes, that would definitely warrant a comment for the scrubs like me that don't know their VBA :wink:

Reply:
That came up in chat the other day. I had it wrong on *my* first go because I originally assumed that VBA restricted you to a single underscore.  I'll see if I can track down a reference in the language spec to comment it with.